pkg/querier/worker/worker.go

```diff
@@ -217,61 +223,130 @@ func (w *querierWorker) AddressAdded(address string) {
 	w.mu.Lock()
 	defer w.mu.Unlock()
 
+	address := instance.Address
 	if m := w.managers[address]; m != nil {
 		return
 	}
 
-	level.Info(w.log).Log("msg", "adding connection", "addr", address)
+	level.Info(w.log).Log("msg", "adding connection", "addr", address, "in-use", instance.InUse)
 	conn, err := w.connect(context.Background(), address)
 	if err != nil {
 		level.Error(w.log).Log("msg", "error connecting", "addr", address, "err", err)
 		return
 	}
 
 	w.managers[address] = newProcessorManager(ctx, w.processor, conn, address)
+	w.instances[address] = instance
+
 	// Called with lock.
 	w.resetConcurrency()
 }
 
-func (w *querierWorker) AddressRemoved(address string) {
-	level.Info(w.log).Log("msg", "removing connection", "addr", address)
+func (w *querierWorker) InstanceRemoved(instance servicediscovery.Instance) {
+	address := instance.Address
+
+	level.Info(w.log).Log("msg", "removing connection", "addr", address, "in-use", instance.InUse)
 
 	w.mu.Lock()
 	p := w.managers[address]
 	delete(w.managers, address)
+	delete(w.instances, address)
 	w.mu.Unlock()
 
 	if p != nil {
 		p.stop()
 	}
+
+	// Re-balance the connections between the available query-frontends / query-schedulers.
+	w.mu.Lock()
+	w.resetConcurrency()
```

Comment:
Note to reviewers: Not calling `resetConcurrency()` was a bug, because it means we don't rebalance the connections when a instance leaves. I've added a dedicated CHANGELOG entry to mention it.

Reply:
Nice find. In typical usage when one scheduler stops, another one also starts and we call resetConcurrency then. But we should not rely on that.